### PR TITLE
expression, plan: improve fold constant

### PR DESCRIPTION
**Pull request by XuHuaiyu**
We do not need to traverse the expression tree every time when folding constant.
```
for condition like:
`a or a or a or a or a or a or a or a or a or a`

BenchmarkFoldConstantBefore-4          100000             17296 ns/op
BenchmarkFoldConstantInAfter-4        100000             14927 ns/op
```
PTAL @winkyao @hanfei1991 @zz-jason

### CODE REVIEW

File: expression/constant_fold.go [hunk 1]
```diff
@@ -18,16 +18,20 @@ import (
 )
 
 // FoldConstant does constant folding optimization on an expression.
-func FoldConstant(expr Expression) Expression {
+// isRecursive indicates whether folding constant in current layer or recursively.
+func FoldConstant(expr Expression, isRecursive bool) Expression {
```
Comment (hanfei1991) [inline expression/constant_fold.go]: We can remove the isRecursive argument.

**Comment by zz-jason** [comment]
LGTM

**Comment by XuHuaiyu** [comment]
PTAL @hanfei1991 
/run-all-test

File: expression/expression.go [hunk 2]
```diff
@@ -290,7 +290,7 @@ func ColumnInfos2ColumnsWithDBName(dbName, tblName model.CIStr, colInfos []*mode
 
 // NewCastFunc creates a new cast function.
 func NewCastFunc(tp *types.FieldType, arg Expression, ctx context.Context) Expression {
-	return FoldConstant(BuildCastFunction(ctx, arg, tp))
+	return BuildCastFunction(ctx, arg, tp)
```
Comment (hanfei1991) [inline expression/expression.go]: What is the difference between `NewCastFunc` and `BuildCastFunction` ??

Comment (XuHuaiyu) [reply expression/expression.go]: Yes, no difference now.

**Comment by XuHuaiyu** [comment]
PTAL @hanfei1991

**Pull request merged**